src/v/cloud_storage/materialized_segments.cc

```diff
@@ -195,54 +256,17 @@ void materialized_segments::trim_segments() {
 
     // These pointers are safe because there are no scheduling points
     // between here and the ultimate eviction at end of function.
-    std::vector<std::pair<materialized_segment_state*, kafka::offset>>
-      to_offload;
-
+    offload_list_t to_offload;
     for (auto& st : _materialized) {
         auto deadline = st.atime + stm_max_idle_time;
-        if (now >= deadline && !st.segment->download_in_progress()) {
-            if (st.segment.owned()) {
-                // This segment is not referred to by any readers, we may
-                // enqueue it for eviction.
-                vlog(
-                  cst_log.debug,
-                  "Materialized segment with base offset {} is stale",
-                  st.offset_key);
-                // this will delete and unlink the object from
-                // _materialized collection
-                if (st.parent) {
-                    to_offload.push_back(std::make_pair(&st, st.offset_key));
-                } else {
-                    // This cannot happen, because materialized_segment_state
-                    // is only instantiated by remote_partition and will
-                    // be disposed before the remote_partition it points to.
-                    vassert(
-                      false,
-                      "materialized_segment_state outlived remote_partition");
-                }
-            } else {
-                // We would like to trim this segment, but cannot right now
-                // because it has some readers referring to it.  Enqueue these
-                // readers for eviction, in the expectation that on the next
-                // periodic pass through this function, the segment will be
-                // eligible for eviction, if it does not create any new readers
-                // in the meantime.
-                vlog(
-                  cst_log.debug,
-                  "Materialized segment {} base-offset {} is not stale: {} "
-                  "readers={}",
-                  st.ntp(),
-                  st.base_rp_offset,
-                  st.segment.use_count(),
-                  st.readers.size());
-
-                // Readers hold a reference to the segment, so for the
-                // segment.owned() check to pass, we need to clear them out.
-                while (!st.readers.empty()) {
-                    evict_reader(std::move(st.readers.front()));
-                    st.readers.pop_front();
-                }
-            }
+
+        // We freed enough, drop out of iterating over segments
+        if (target_free && _reader_units.current() >= target_free) {
```

Comment:
We should check against `_segment_units` here, no?

Reply:
good catch, fixed.